File: src/PoMiniGames.Client/src/games/connectfive/ConnectFiveAI.ts
import { Piece, Difficulty } from '../shared/types';
import { ConnectFiveBoard } from './ConnectFiveBoard';

/**
 * Connect Five AI with three difficulty levels.
 * - Easy: random valid column
 * - Medium: win → block → centre bias → random
 * - Hard: minimax with alpha-beta pruning (depth 5)
 */
export class ConnectFiveAI {
  static getMove(board: ConnectFiveBoard, player: Piece, difficulty: Difficulty): number {
    switch (difficulty) {
      case Difficulty.Easy:
        return ConnectFiveAI.easyMove(board);
      case Difficulty.Medium:
        return ConnectFiveAI.mediumMove(board, player);
      case Difficulty.Hard:
        return ConnectFiveAI.hardMove(board, player);
    }
  }

  // ─── Easy ──────────────────────────────────────────────────────
  private static easyMove(board: ConnectFiveBoard): number {
    const cols = board.getAvailableColumns();
    return cols[Math.floor(Math.random() * cols.length)]!;
  }

  // ─── Medium ────────────────────────────────────────────────────
  private static mediumMove(board: ConnectFiveBoard, player: Piece): number {
    const opponent = player === Piece.Red ? Piece.Yellow : Piece.Red;

    // 1. Win if possible
    const win = ConnectFiveAI.findWinningCol(board, player);
    if (win !== null) return win;

    // 2. Block opponent win
    const block = ConnectFiveAI.findWinningCol(board, opponent);
    if (block !== null) return block;

    // 3. Centre bias
    const mid = Math.floor(ConnectFiveBoard.Cols / 2);
    if (board.getTargetRow(mid) >= 0) return mid;

    // 4. Random
    return ConnectFiveAI.easyMove(board);
  }

  // ─── Hard (minimax) ───────────────────────────────────────────
  private static readonly MAX_DEPTH = 5;

  private static hardMove(board: ConnectFiveBoard, player: Piece): number {
    const opponent = player === Piece.Red ? Piece.Yellow : Piece.Red;
    let bestScore = -Infinity;
    let bestCol = board.getAvailableColumns()[0]!;

    // Evaluate columns from centre outward for better pruning
    const cols = ConnectFiveAI.centreOrdered(board.getAvailableColumns());

    for (const col of cols) {
      const next = board.drop(col, player);
      const score = ConnectFiveAI.minimax(next, ConnectFiveAI.MAX_DEPTH - 1, -Infinity, Infinity, false, player, opponent);
      if (score > bestScore) {
        bestScore = score;
        bestCol = col;
      }
    }
    return bestCol;
  }

  private static minimax(
    board: ConnectFiveBoard,
    depth: number,
    alpha: number,
    beta: number,
    isMaximising: boolean,
    player: Piece,
    opponent: Piece,
  ): number {
    if (board.checkWin(player).won) return 100000 + depth;
    if (board.checkWin(opponent).won) return -(100000 + depth);
    if (board.isFull() || depth === 0) return ConnectFiveAI.heuristic(board, player, opponent);

    const cols = ConnectFiveAI.centreOrdered(board.getAvailableColumns());

    if (isMaximising) {
      let best = -Infinity;
      for (const col of cols) {
        const next = board.drop(col, player);
        const score = ConnectFiveAI.minimax(next, depth - 1, alpha, beta, false, player, opponent);
        best = Math.max(best, score);
        alpha = Math.max(alpha, best);
        if (beta <= alpha) break;
      }
      return best;
    } else {
      let best = Infinity;
      for (const col of cols) {
        const next = board.drop(col, opponent);
        const score = ConnectFiveAI.minimax(next, depth - 1, alpha, beta, true, player, opponent);
        best = Math.min(best, score);
        beta = Math.min(beta, best);
        if (beta <= alpha) break;
      }
      return best;
    }
  }

  private static heuristic(board: ConnectFiveBoard, player: Piece, opponent: Piece): number {
    let score = 0;
    const R = ConnectFiveBoard.Rows;
    const C = ConnectFiveBoard.Cols;
    const W = ConnectFiveBoard.WinLength;
    const directions = [[0, 1], [1, 0], [1, 1], [1, -1]] as const;

    for (let r = 0; r < R; r++) {
      for (let c = 0; c < C; c++) {
        for (const [dr, dc] of directions) {
          let mine = 0;
          let theirs = 0;
          let empty = 0;
          let inBounds = true;
          for (let i = 0; i < W; i++) {
            const nr = r + dr * i;
            const nc = c + dc * i;
            if (nr < 0 || nr >= R || nc < 0 || nc >= C) { inBounds = false; break; }
            const cell = board.get(nr, nc);
            if (cell === player) mine++;
            else if (cell === opponent) theirs++;
            else empty++;
          }
          if (!inBounds) continue;
          if (theirs === 0 && mine > 0) score += ConnectFiveAI.windowScore(mine);
          if (mine === 0 && theirs > 0) score -= ConnectFiveAI.windowScore(theirs);
        }
      }
    }

    // Centre-column bonus
    const midCol = Math.floor(C / 2);
    for (let r = 0; r < R; r++) {
      if (board.get(r, midCol) === player) score += 3;
      if (board.get(r, midCol) === opponent) score -= 3;
    }

    return score;
  }

  private static windowScore(count: number): number {
    switch (count) {
      case 4: return 100;
      case 3: return 10;
      case 2: return 3;
      case 1: return 1;
      default: return 0;
    }
  }

  private static centreOrdered(cols: number[]): number[] {
    const mid = Math.floor(ConnectFiveBoard.Cols / 2);
    return [...cols].sort((a, b) => Math.abs(a - mid) - Math.abs(b - mid));
  }

  private static findWinningCol(board: ConnectFiveBoard, player: Piece): number | null {
    for (const col of board.getAvailableColumns()) {
      const next = board.drop(col, player);
      if (next.checkWin(player).won) return col;
    }
    return null;
  }
}

File: src/PoMiniGames.Client/Games/ConnectFive/ConnectFiveAI.cs
using PoMiniGamesClient.Enums;
using PoMiniGamesClient.Models;

namespace PoMiniGamesClient.Games.ConnectFive;

public class ConnectFiveAI
{
    public (int Row, int Col) GetMove(ConnectFiveBoard board, Player player, Difficulty difficulty)
    {
        return difficulty switch
        {
            Difficulty.Easy   => EasyMove(board, player),
            Difficulty.Medium => MediumMove(board, player),
            Difficulty.Hard   => HardMove(board, player),
            _                 => MediumMove(board, player)
        };
    }

    /// <summary>
    /// Pick a move whose strength scales with <paramref name="cpuElo"/>. The ELO
    /// maps to a (search depth, blunder-rate) rung: at a low rating the CPU mostly
    /// plays random moves (missing blocks and wins); as the rating climbs it plays
    /// deeper negamax with fewer blunders. This lets the 1-player game match the
    /// CPU to the player's adaptive rating for a ~50/50 challenge.
    /// </summary>
    public (int Row, int Col) GetAdaptiveMove(ConnectFiveBoard board, Player player, int cpuElo)
    {
        var (depth, blunder) = StrengthForElo(cpuElo);

        // A "blunder" is a fully random legal move — it may miss its own win or
        // fail to block the player, which is exactly what makes low rungs weak.
        if (blunder > 0 && Random.Shared.NextDouble() < blunder)
        {
            var cols = board.GetAvailableCols();
            if (cols.Count == 0) return (0, 0);
            var randCol = cols[Random.Shared.Next(cols.Count)];
            return (board.GetTargetRow(randCol), randCol);
        }

        // Otherwise play a "smart" move: depth 1 = the win/block/center heuristic,
        // depth >= 2 = alpha-beta negamax to that depth.
        if (depth <= 1) return MediumMove(board, player);

        var search = new NegamaxSearch(player, maxDepth: depth);
        return search.ChooseMove(board);
    }

    /// <summary>
    /// Maps a CPU ELO to a strength rung. Seven rungs span roughly 800–2000; the
    /// top rung is the existing 4-ply "Hard" search with no blunders. Depth is
    /// capped at 4 to keep every move well inside the game's responsiveness budget.
    /// </summary>
    public static (int Depth, double Blunder) StrengthForElo(int elo) => elo switch
    {
        < 900  => (1, 0.60),
        < 1050 => (1, 0.38),
        < 1200 => (2, 0.24),
        < 1350 => (2, 0.12),
        < 1500 => (3, 0.06),
        < 1700 => (4, 0.03),
        _      => (4, 0.00),
    };

    private (int, int) EasyMove(ConnectFiveBoard board, Player player)
    {
        var opponent = player.Other;

        if (Random.Shared.NextDouble() < 0.3)
        {
            var block = FindWinningMove(board, opponent);
            if (block.HasValue) return block.Value;
        }

        var cols = board.GetAvailableCols();
        if (cols.Count == 0) return (0, 0);
        var col = cols[Random.Shared.Next(cols.Count)];
        return (board.GetTargetRow(col), col);
    }

    private (int, int) MediumMove(ConnectFiveBoard board, Player player)
    {
        var opponent = player.Other;

        var win = FindWinningMove(board, player);
        if (win.HasValue) return win.Value;

        var block = FindWinningMove(board, opponent);
        if (block.HasValue) return block.Value;

        var cols = board.GetAvailableCols();
        if (cols.Count == 0) return (0, 0);

        var centerCol = ConnectFiveBoard.Cols / 2;
        if (cols.Contains(centerCol))
        {
            return (board.GetTargetRow(centerCol), centerCol);
        }

        // Bug fix 2026-07-05: the previous return used a *random* column for the
        // row lookup but *cols[0]* for the column returned. If the random and the
        // fixed index differed (e.g. cols=[0,1,2], random picks 2, return col 0),
        // the row was the bottom of col 2 but the place went into col 0 — a
        // piece could land above the existing stack in col 0. The fix: pick one
        // column and use its target row for both halves of the tuple.
        var pick = cols[Random.Shared.Next(cols.Count)];
        return (board.GetTargetRow(pick), pick);
    }

    private (int, int) HardMove(ConnectFiveBoard board, Player player)
    {
        // Audit #7: 4-ply negamax with alpha-beta + a small transposition table.
        // Bounded by the existing AiMoveBudgetMs in RunAiTurnAsync (8 s ceiling).
        // Medium difficulty calls the same search at depth=2 to stay sub-100ms.
        var search = new NegamaxSearch(player, maxDepth: 4);
        return search.ChooseMove(board);
    }

    private (int, int)? FindWinningMove(ConnectFiveBoard board, Player player)
    {
        foreach (var col in board.GetAvailableCols())
        {
            var row = board.GetTargetRow(col);
            var next = board.Place(row, col, player);
            if (next.CheckWin(player).Won) return (row, col);
        }
        return null;
    }
}